ntoskrnl/inbv/inbv.c

```diff
@@ -932,18 +951,147 @@ DisplayBootBitmap(IN BOOLEAN TextMode)
+        InbvRotBarInit();
     }
 #endif
 }
 
+#ifdef INBV_ROTBAR_IMPLEMENTED
+static
+VOID
+NTAPI
+InbvRotationThread(
+    _In_ PVOID Context)
+{
+    ULONG X, Y, Index = 0, Total = 18;
+
+    InbvAcquireLock();
+    X = ProgressBarLeft + 2;
+    Y = ProgressBarTop + 2;
+    InbvReleaseLock();
+
+    while (TRUE)
+    {
+        InbvAcquireLock();
+
+        if (InbvDisplayState != INBV_DISPLAY_STATE_OWNED || RotBarSelection != RB_UNSPECIFIED)
+        {
+            InbvReleaseLock();
+            break;
+        }
+
+        if (Index >= 3)
+        {
+            /* Fill previous bar position */
+            VidSolidColorFill(X + ((Index - 3) * 8), Y, (X + ((Index - 3) * 8)) + 8 - 1, Y + 9 - 1, 0);
+        }
+        if (Index < Total - 1)
+        {
+            /* Draw the progress bar bit */
+            if (Index < 2)
+            {
+                /* Appearing from the left */
+                VidBufferToScreenBlt(RotBarBuffer + 8 * (2 - Index), X, Y, 22 - 8 * (2 - Index), 9, 24);
+            }
+            else if (Index >= Total - 3)
+            {
+                /* Hiding to the right */
+                VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22 - 8 * (4 - (Total - Index)), 9, 24);
+            }
+            else
+            {
+                VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22, 9, 24);
+            }
+        }
+
+        InbvReleaseLock();
+
+        /* Wait for a bit */
+        KeStallExecutionProcessor(ROT_BAR_TIME);
+
+        Index = (Index + 1) % Total;
+    }
+    PsTerminateSystemThread(STATUS_SUCCESS);
+}
+
+VOID
+NTAPI
+INIT_FUNCTION
+InbvRotBarInit(VOID)
+{
+    NTSTATUS Status;
+    HANDLE ThreadHandle = NULL;
+
+    if (!InbvBootDriverInstalled || RotBarSelection == RB_UNSPECIFIED)
+        return;
+
+    RotBarThread = NULL;
+
+    Status = PsCreateSystemThread(&ThreadHandle,
+                                  SYNCHRONIZE,
+                                  NULL,
+                                  NULL,
+                                  NULL,
+                                  InbvRotationThread,
+                                  NULL);
+
+    if (NT_SUCCESS(Status) && ThreadHandle != NULL)
+    {
+        Status = ObReferenceObjectByHandle(ThreadHandle,
+                                           SYNCHRONIZE,
+                                           PsThreadType,
+                                           KernelMode,
+                                           &RotBarThread,
+                                           NULL);
+        if (NT_SUCCESS(Status) && RotBarThread)
+            RotBarThreadActive = TRUE;
+        ObCloseHandle(ThreadHandle, KernelMode);
+    }
+}
+
+VOID
+NTAPI
+INIT_FUNCTION
+InbvRotBarStop(VOID)
+{
+    NTSTATUS Status;
+
+    if (!RotBarThread)
+    {
+        RotBarThreadActive = FALSE;
+        return;
+    }
+
+    RotBarSelection = RB_UNSPECIFIED;
+
+    Status = KeWaitForSingleObject(RotBarThread,
```

Comment:
By the way, if your code gives a reasonable guarantee that the thread will terminate on its own, you don't need this wait at all, or to keep track of the thread pointer.

Reply:
The animation thread needs to be controlled outside, because the screen goes black a moment before the Inbv looses its `INBV_DISPLAY_STATE_OWNED` state.

Reply:
All you seem to do is wait for it to terminate, I don't see any other "control" happening. Is it really necessary to wait for the thread here before other operations can continue? You don't need to wait just because it seems cleaner or something.

Reply:
> Is it really necessary to wait for the thread here before other operations can continue?

Yes, it's necessary because in some cases it can draw things [after screen fill](https://github.com/reactos/reactos/pull/194/files#diff-0329f9717fc2c145477c134367a995c9R1167) (tested).

Reply:
Right, because you have a race condition in there. I'll comment in the corresponding code.